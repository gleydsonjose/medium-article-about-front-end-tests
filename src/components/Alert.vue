<template>
  <div class="alert" :style="alertStyle">
    <section class="alert__message-outer">
      <i
        v-show="alert.useIcon"
        class="mdi alert__type-icon"
        :class="colorsType.icon"
      ></i>
      <p class="alert__message">{{ alert.message }}</p>
    </section>
    <button class="alert__btn-close" @click="close">
      <i class="mdi mdi-close-circle alert__btn-close-icon"></i>
    </button>
  </div>
</template>

<script>
import { COLORS_TYPE } from "@/constants/colors.js"
import { mapState, mapActions } from "vuex";

export default {
  name: "Alert",
  computed: {
    ...mapState(["alert"]),
    colorsType() {
      const type = this.alert?.type || "default";
      return COLORS_TYPE[type];
    },
    alertStyle() {
      const style = {
        "background-color": this.colorsType.backgroundColor,
        "border-color": this.colorsType.borderColor,
      };

      return style;
    },
  },
  methods: {
    ...mapActions(["closeAlert"]),
    close() {
      this.closeAlert();
    },
  },
};
</script>

<style scoped>
.alert {
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  min-height: 40px;
  border-style: solid;
  border-width: 1px;
  border-radius: 5px;
  color: #eceff1;
  padding: 8px 0;
  font-size: 12pt;
}

.alert__message-outer {
  display: flex;
  align-items: center;
}

.alert__type-icon {
  margin: 0 6px;
}

.alert__message {
  word-wrap: break-word;
  width: 396px;
  font-weight: bold;
  margin: 0;
}

.alert__btn-close {
  background: transparent;
  border: none;
  color: #eceff1;
  padding: 0;
  transition: opacity 0.25s ease-out;
  margin: 0 6px;
}

.alert__btn-close:hover {
  opacity: 0.6;
  cursor: pointer;
  transition: opacity 0.25s ease-in;
}

.alert__btn-close-icon {
  font-size: 13pt;
}
</style>
