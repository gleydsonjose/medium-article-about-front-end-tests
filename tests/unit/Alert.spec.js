import { createLocalVue, shallowMount } from "@vue/test-utils";
import Alert from "@/components/Alert.vue";
import Vuex from "vuex";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("<template/>", () => {
  let state;
  let actions;

  beforeEach(() => {
    state = {
      alert: {
        isOpen: true,
        type: "default",
        message: "",
        useIcon: false,
      },
    };
    actions = {
      closeAlert: () => { },
    };
  });

  it("deveria mostrar o icone após usar prop useIcon", () => {
    state.alert.useIcon = true;

    const wrapper = shallowMount(Alert, {
      store: new Vuex.Store({
        state,
        actions,
      }),
      localVue,
    });

    expect(wrapper.find(".alert__type-icon").isVisible()).toBeTruthy();
  });

  it("deveria esconder o icone após não usar prop useIcon", () => {
    const wrapper = shallowMount(Alert, {
      store: new Vuex.Store({
        state,
        actions,
      }),
      localVue,
    });

    expect(wrapper.find(".alert__type-icon").isVisible()).toBeFalsy();
  });

  it("deveria acionar evento após clicar no botão de fechar alerta", () => {
    const closeMethodSpy = jest.spyOn(Alert.methods, "close");
    const wrapper = shallowMount(Alert, {
      store: new Vuex.Store({
        state,
        actions,
      }),
      localVue,
    });

    wrapper.find(".alert__btn-close").trigger("click");

    expect(closeMethodSpy).toBeCalled();
  });

  it("deveria ser igual ao snapshot", () => {
    const wrapper = shallowMount(Alert, {
      store: new Vuex.Store({
        state,
        actions,
      }),
      localVue,
    });

    expect(wrapper).toMatchSnapshot();
  });
});

describe("methods@close", () => {
  let state;
  let actions;

  beforeEach(() => {
    state = {
      alert: {
        isOpen: true,
        type: "default",
        message: "",
        useIcon: false,
      },
    };
    actions = {
      closeAlert: () => { },
    };
  });

  it("deveria chamar action do Vuex após usar método", () => {
    const closeAlert = jest.fn();
    actions.closeAlert = closeAlert;
    const wrapper = shallowMount(Alert, {
      store: new Vuex.Store({
        state,
        actions,
      }),
      localVue,
    });

    wrapper.find(".alert__btn-close").trigger("click");

    expect(closeAlert).toBeCalled();
  });
});
