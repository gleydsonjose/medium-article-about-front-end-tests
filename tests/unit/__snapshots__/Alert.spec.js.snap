// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<template/> deveria ser igual ao snapshot 1`] = `
<div class="alert" style="background-color: rgb(55, 71, 79); border-color: #263238;">
  <section class="alert__message-outer"><i class="mdi alert__type-icon mdi-alert-circle" style="display: none;"></i>
    <p class="alert__message"></p>
  </section> <button class="alert__btn-close"><i class="mdi mdi-close-circle alert__btn-close-icon"></i></button>
</div>
`;
